unstructured/partition/code.py

```diff
@@ -0,0 +1,226 @@
+        and (min_partition > max_partition or min_partition < 0 or max_partition < 0)
+    ):
+        raise ValueError("Invalid values for min_partition and/or max_partition.")
+
+    # Verify that only one of the arguments was provided
+    exactly_one(filename=filename, file=file, text=text)
+    file_text = bytes()
+    programming_language = None
+    last_modification_date = None
+
+    if filename is not None:
+        try:
+            with open(filename, "rb") as f:
+                file_text = f.read()
+            _, extension = os.path.splitext(filename)
+            filetype = EXT_TO_FILETYPE.get(extension)
+            if filetype:
+                programming_language = FILETYPE_TO_LANG.get(filetype)
+            last_modification_date = get_last_modified_date(filename)
+        except FileNotFoundError:
+            raise FileNotFoundError("Provided filename is not correct")
+    elif text is not None:
+        file = io.BytesIO(bytes(text, "utf-8"))
+
+    if file is not None:
+        filetype = detect_filetype(file=file)
+        file.seek(0)
+        file_text = file.read()
+        last_modification_date = get_last_modified_date_from_file(file)
+        if filetype:
+            programming_language = FILETYPE_TO_LANG.get(filetype)
+        else:
+            raise RuntimeError("Unable to detect code file type")
+
+    if min_partition is not None and len(file_text) < min_partition:
+        min_partition = len(file_text)
+        # raise ValueError("`min_partition` cannot be larger than the length of file contents.")
+
+    if programming_language is None:
+        raise ValueError("No programming language provided/detected")
+
+    global TREE_SITTER_BUILD_PATH
+    try:
+        LANG = tree_sitter.Language(TREE_SITTER_BUILD_PATH, name=programming_language)
+    except (AttributeError, OSError):
+        update_tree_sitter([programming_language])
+        LANG = tree_sitter.Language(TREE_SITTER_BUILD_PATH, name=programming_language)
+
+    parser = tree_sitter.Parser()
+    parser.set_language(LANG)
+    tree = parser.parse(file_text)
+    if not tree.root_node.children or tree.root_node.children[0].type == "ERROR":
+        raise ValueError(f"File is not written in {programming_language}")
+
+    file_content, _ = _partition_by_node(
+        node=tree.root_node,
+        text=file_text.decode("utf-8"),
+        min_chars=min_partition,
+        max_chars=max_partition,
+    )
+
+    elements: List[Element] = []
+
+    if include_metadata:
+        metadata = ElementMetadata(
+            filename=metadata_filename or filename,
+            last_modified=metadata_last_modified or last_modification_date,
+            languages=[programming_language],
+        )
+        metadata.detection_origin = detection_origin
+    else:
+        metadata = ElementMetadata()
+
+    # Note (Pierre) Some languages (i.e Python) use blank space and tabs for semantic
+    # Maybe removing the spaces is not a good idea
```

Comment:
I agree to some extent.
In a language that enforces the use of brackets for inner blocks, we can then retrieve the actual context by adding the chunks before and after.
For Python (and maybe other languages but I don't know), there is no way to understand where a code block fits unless we have it's indentation (or check for errors)

We could re-format all code coming in and then find a more compact representation for the indentation, but it seems a bit too involved

Reply:
Hmm, @Plemeur I think I see the complication you indicate. When we _combine_ elements (concatenate their text), if we've gotten rid of the indentation at the TreeSitter (TS) node level, they could combine to produce invalid code, at least in Python.

So if we did decide to do this, it would have to be at the element level (encapsulated in `CodeElementBuilder`) such that "leading" indentation was removed consistently for all nodes in a single element.

Also, we would have to expect that elements were never combined during chunking.

The more I think about it the more the current approach seems to be partitioning and chunking welded together, which makes me uneasy about it fitting in consistently with the broader pipeline concepts across other document types.

At the same time it's not clear to me what an "atomic" element would be. Clearly a function would be too big, at least sometimes, and a statement would be the next level down; I don't think we could distinguish "passages" without some sort of inference.

@Plemeur does the division this produce turn out to be effective in your application? And would there be the prospect of chunking these elements with one of the unstructured chunkers?

Reply:
I'm just starting out with this
I'm using this https://github.com/bionic-gpt/bionic-gpt and for now and my plan was to go up the chain from the chunking to the agent to understand how things work together

From what I understand from the chunkers code, it would not really be applicable to code broadly because of the different writing styles

I don't feel there is a valid abstraction that will satisfy 300 lines of parameters definition in a python dataclass, and at the same time 300 lines of JSX in the return of a react functional component

For the "atomic" element talk, when I tried to rewrite the parsing the first time, I was trying to find a good logic but came out to those conclusion: Any element might be too long to be considered atomic. 
Even if you think "Well a value declaration should be small enough" ... Hard coded dictionary, list, or strings longer than function will appear, and there's nothing you can do about it

Reply:
I'm okay with keeping the indent in the elements for now, even if it creates some bloat. We do have #711 out there to add a metadata field for indentation. At that point, we could add the indentation level of the code as metadata and skip the extra test.

Reply:
@Plemeur Yeah, on reflection I think you're right about the lack of a naturally-occurring "atomic" element in code. After a day's reflection I can't think of any uniform structure that would always be guaranteed to be shorter than any given threshold.

Also I'm inclined to agree that code is distinctive enough from prose/narrative text that it's not going to be compatible in-general with chunkers that are really designed for narrative text. In particular, there are no `Title` elements in the stream so `chunk_by_title()` wouldn't be applicable and there are no `Table` elements, so a good portion of the chunking code would never get triggered. My working conclusion from that is that if we do have code-element chunkers they would need to be dedicated to code element-streams.

Along the lines of brainstorming, it occurs to me we might one day want elements that as much as possible are whole functions or perhaps [Module | Class | Function]. Like other elements, these would not have a size limit. Then separately a chunker would divide oversized functions into sub-chunks, possibly by a second treesitter parse. Just an idea, but does argue to me for keeping chunking distinct from partitioning, as we do for narrative text. 

Right now I'd say we do have partitioning and chunking, even if they're "welded together" in this implementation:
- Partitioning is generating whole TS nodes that fit in the max_char window.
- Chunking is combining those to be between min and max-chars.

Maybe we'll dream up alternate chunking methods going forward, which would be an argument for keeping these two operations distinct such that they can be flexibly composed into other configurations.

Have you given any thought to the metadata that might make sense for code chunks? I'm inclined to think that hierarchy is readily available and some of that might be worth recording at some point.

All this isn't to do with what your code here per se, just noodling on how code partitioning might evolve so we can preserve flexibilities we might want as it evolves.

Reply:
I understand why you want to keep them separate.

I did at first try to partition more sparsely walking the treesitter tree, grouping elements that have similar nature together if they are close enough, but I had some issues making sense of child elements at times.

For instance 
```python
def LongFunction(
    param_1,
    param_2
   ) :
   function body ....
```

When you arrive to the Long function element, the children nodes are `def` `LongFunction` `(` `param_1` `param_2` `):` `body`. 
So it requires some heavy logic to put everything together properly (and is most likely language dependent)

Reply:
Hmm, yes, that's a good point. I believe the node types from TreeSitter are language-specific, right? So there's no such thing as a "function" node that is the same for all languages. So we would have to maintain some mapping between language and node-type names, which would be perpetually incomplete. Hmm, yes, that would definitely support the idea of not getting too specific with trying to distinguish node types. Glad you mentioned that :)